File: backend/app/game/state.py
"""Board state representation and utilities."""
from enum import Enum
from typing import Optional
from pydantic import BaseModel
import copy


class Player(str, Enum):
    RED = "red"
    WHITE = "white"


class PieceType(str, Enum):
    MAN = "man"
    KING = "king"


class Piece(BaseModel):
    player: Player
    type: PieceType


class GameState(BaseModel):
    board: list[list[Optional[Piece]]]
    current_player: Player
    move_history: list[dict]
    captured_pieces: dict[str, int]
    game_over: bool
    winner: Optional[str]


def create_initial_state() -> GameState:
    """Create initial board state with pieces in starting positions."""
    board = [[None for _ in range(8)] for _ in range(8)]

    # Place red pieces (top 3 rows)
    for row in range(3):
        for col in range(8):
            if (row + col) % 2 == 1:
                board[row][col] = Piece(player=Player.RED, type=PieceType.MAN)

    # Place white pieces (bottom 3 rows)
    for row in range(5, 8):
        for col in range(8):
            if (row + col) % 2 == 1:
                board[row][col] = Piece(player=Player.WHITE, type=PieceType.MAN)

    return GameState(
        board=board,
        current_player=Player.RED,
        move_history=[],
        captured_pieces={Player.RED.value: 0, Player.WHITE.value: 0},
        game_over=False,
        winner=None,
    )


def clone_state(state: GameState) -> GameState:
    """Deep clone a state for AI simulation."""
    return GameState(
        board=[[piece.model_copy() if piece else None for piece in row] for row in state.board],
        current_player=state.current_player,
        move_history=[m.copy() for m in state.move_history],
        captured_pieces=state.captured_pieces.copy(),
        game_over=state.game_over,
        winner=state.winner,
    )


def get_piece(state: GameState, row: int, col: int) -> Optional[Piece]:
    """Get piece at position."""
    if not (0 <= row <= 7 and 0 <= col <= 7):
        return None
    return state.board[row][col]


def is_dark_square(row: int, col: int) -> bool:
    """Check if position is valid dark square."""
    return (row + col) % 2 == 1


def count_pieces(state: GameState, player: Player) -> dict:
    """Count pieces for a player."""
    total = 0
    kings = 0
    for row in range(8):
        for col in range(8):
            piece = state.board[row][col]
            if piece and piece.player == player:
                total += 1
                if piece.type == PieceType.KING:
                    kings += 1
    return {"total": total, "kings": kings, "men": total - kings}


def get_opponent(player: Player) -> Player:
    """Get opponent player."""
    return Player.WHITE if player == Player.RED else Player.RED


def has_valid_moves(state: GameState, player: Player) -> bool:
    """Check if a player has any valid moves."""
    from .moves import generate_legal_moves

    # Temporarily set current player to check their moves
    original_player = state.current_player
    state.current_player = player
    moves = generate_legal_moves(state)
    state.current_player = original_player

    return len(moves) > 0


def get_winner(state: GameState) -> Optional[str]:
    """Check if there's a winner."""
    red_count = count_pieces(state, Player.RED)
    white_count = count_pieces(state, Player.WHITE)

    # Check if a player has no pieces
    if red_count["total"] == 0:
        return Player.WHITE.value
    if white_count["total"] == 0:
        return Player.RED.value

    # Check if current player has no valid moves (stalemate)
    if not state.game_over:
        if not has_valid_moves(state, state.current_player):
            # Current player loses, opponent wins
            return get_opponent(state.current_player).value

    return None

File: frontend/src/game/engine/state.js
// Board state representation and utilities

export const PLAYERS = {
  RED: 'red',
  WHITE: 'white',
};

export const PIECE_TYPES = {
  MAN: 'man',
  KING: 'king',
};

/**
 * Create initial board state
 * Red pieces on top (rows 0-2), White pieces on bottom (rows 5-7)
 * Only dark squares have pieces
 */
export function createInitialState() {
  const board = Array(8).fill(null).map(() => Array(8).fill(null));

  // Place red pieces (top 3 rows)
  for (let row = 0; row < 3; row++) {
    for (let col = 0; col < 8; col++) {
      if ((row + col) % 2 === 1) {
        board[row][col] = { player: PLAYERS.RED, type: PIECE_TYPES.MAN };
      }
    }
  }

  // Place white pieces (bottom 3 rows)
  for (let row = 5; row < 8; row++) {
    for (let col = 0; col < 8; col++) {
      if ((row + col) % 2 === 1) {
        board[row][col] = { player: PLAYERS.WHITE, type: PIECE_TYPES.MAN };
      }
    }
  }

  return {
    board,
    currentPlayer: PLAYERS.RED, // Red moves first
    moveHistory: [],
    capturedPieces: { [PLAYERS.RED]: 0, [PLAYERS.WHITE]: 0 },
    gameOver: false,
    winner: null,
  };
}

/**
 * Deep clone a state for AI simulation
 */
export function cloneState(state) {
  return {
    board: state.board.map(row => row.map(cell => cell ? { ...cell } : null)),
    currentPlayer: state.currentPlayer,
    moveHistory: [...state.moveHistory],
    capturedPieces: { ...state.capturedPieces },
    gameOver: state.gameOver,
    winner: state.winner,
  };
}

/**
 * Get piece at position
 */
export function getPiece(state, row, col) {
  if (row < 0 || row > 7 || col < 0 || col > 7) return null;
  return state.board[row][col];
}

/**
 * Check if position is valid dark square
 */
export function isDarkSquare(row, col) {
  return (row + col) % 2 === 1;
}

/**
 * Count pieces for a player
 */
export function countPieces(state, player) {
  let count = 0;
  let kings = 0;
  for (let row = 0; row < 8; row++) {
    for (let col = 0; col < 8; col++) {
      const piece = state.board[row][col];
      if (piece && piece.player === player) {
        count++;
        if (piece.type === PIECE_TYPES.KING) kings++;
      }
    }
  }
  return { total: count, kings, men: count - kings };
}

/**
 * Get opponent
 */
export function getOpponent(player) {
  return player === PLAYERS.RED ? PLAYERS.WHITE : PLAYERS.RED;
}
